Use self-hosted web api

SzymonPobiega opened the pull request:
Closes #23

SzymonPobiega commented:
I've changed only one project in one sln to verify the approach. Please review.

adamralph commented on exercises/01-composite-ui/before/01-composite-ui-before.sln:
```diff
@@ -129,6 +128,7 @@ Global
+		{FDD10539-A381-48BD-AA56-328DD9991B9F} = {808B901F-6EF7-49F0-B668-F54EEDB1F3F1}
 	EndGlobalSection
 	GlobalSection(ExtensibilityGlobals) = postSolution
 		SolutionGuid = {FC27BA8B-135F-4456-92CC-AE23DD9989AB}
```
Not so relevant right now, but it seems we should revert all changes to this file before merging.

adamralph commented on exercises/01-composite-ui/before/Divergent.Customers.API/Divergent.Customers.API.csproj:
```diff
@@ -1,232 +1,113 @@
-    </NuGetPackageImportStamp>
-    <TargetFrameworkProfile />
-    <Use64BitIISExpress />
+    <TargetFrameworkVersion>v4.6.2</TargetFrameworkVersion>
```
Is the target framework change necessary?

SzymonPobiega replied:
My mistake

adamralph commented on exercises/01-composite-ui/before/Divergent.Customers.API/Program.cs:
```diff
@@ -0,0 +1,18 @@
+        {
+            using (WebApp.Start<Startup>("http://localhost:20186"))
+            {
+                Console.WriteLine("Dicergent.Customers.API web server is running.");
```
Typo: "Dicergent"

adamralph commented on exercises/01-composite-ui/before/Divergent.Customers.API/Program.cs:
```diff
@@ -0,0 +1,18 @@
+            {
+                Console.WriteLine("Dicergent.Customers.API web server is running.");
+                Console.WriteLine("Press <enter> to quit.");
+                Console.ReadLine();
```
I'm not sure this is best. The NSB endpoints in the later exercises require <kbd>Ctrl</kbd>+<kbd>C</kbd> which I think is better, to avoid accidental shutdowns:

https://github.com/Particular/Workshop/blob/9db8434954ee388a47cccad7d3520e251e06fb25/exercises/02-publish-subscribe/after/Divergent.Customers/Program.cs#L28-L31

SzymonPobiega commented:
@adamralph I pushed the changes

SzymonPobiega commented:
@adamralph can you re-review? Do we need to involve others to make sure that the model I used can be used for all the excercises? @DavidBoike maybe we can use you as a third pair of :eyes: ?

adamralph commented on exercises/01-composite-ui/before/Divergent.Customers.API/Program.cs:
```diff
@@ -0,0 +1,23 @@
+
+            using (WebApp.Start<Startup>("http://localhost:20186"))
+            {
+                await Console.Out.WriteLineAsync("Divergent.Customers.API web server is running.");
```
Minor point, but I'm not sure it's worth repeating the namespace here, since the console window will already contain the name of the app. Or if we do, it would be better read it from metadata to save having to repeat it here. It could also be good to put the listening URL into a variable and print it to the console. E.g. `Divergent.Customers.API listening at http://localhost:20186`

SzymonPobiega commented:
@adamralph I rebuilt it from scratch i.e. I created a new console app project and removed the old one. I also cursed the "old' project system a lot ;-)

SzymonPobiega commented:
I've pushed a bunch of commits that transform the entire `before` of the first exercise. Please ACK that part before I can move to after.

adamralph commented on exercises/01-composite-ui/before/Divergent.Finance.API/Startup.cs:
```diff
@@ -1,16 +1,18 @@
-[assembly: OwinStartup(typeof(Divergent.Finance.API.Startup))]
 
-namespace Divergent.Finance.API
+namespace Divergent.Customers.API
```
Should still be `Finance`.

adamralph commented on exercises/01-composite-ui/before/Divergent.Customers.API/packages.config:
```diff
@@ -1,28 +1,20 @@
+  <package id="Microsoft.Owin.SelfHost" version="3.1.0" targetFramework="net461" />
   <package id="Microsoft.Web.Infrastructure" version="1.0.0.0" targetFramework="net461" />
-  <package id="Newtonsoft.Json" version="10.0.3" targetFramework="net461" />
+  <package id="Newtonsoft.Json" version="6.0.4" targetFramework="net461" />
```
Is reverting back to this old version intended?

SzymonPobiega replied:
No. I just reverted it/upgraded json to 10 in all API projects.

SzymonPobiega commented:
OK. Moving on to the after sln of ex 1.

SzymonPobiega commented:
We use `Windsor` container in the API hosts in order to pass the endpoint instance to the controller. Do we want to continue doing this? It seems like an overkill.

mauroservienti commented:
If the only option we're left with is a static field somewhere, I vote for a DI container 😄 
Jokes aside, as soon as we'll migrate to NET Core for back-ends as well we can use the built-in DI container, in the meantime Windsor is good enough 😉

adamralph commented:
Generally, this PR should do the minimum amount required to swap the API browser tabs for console windows.

SzymonPobiega commented:
Ok, I pushed the Ex 2 before. The most interesting part is the Sales API. I tried to keep it as close to the original as possible but I could not get the Radical bootstrapper to work. For some reason it does not detect the controller and I had to register it manually. Anybody can help me find what is missing?

SzymonPobiega commented:
Thanks @mauroservienti . I love you :heart:

SzymonPobiega commented:
OK. Now please review the current version of Ex 2 before. When OK I'll apply the same changes to all the other exercises.

adamralph commented on exercises/02-publish-subscribe/before/Divergent.Customers.API/packages.config:
```diff
@@ -1,28 +1,20 @@
+  <package id="Microsoft.Owin.SelfHost" version="3.1.0" targetFramework="net461" />
   <package id="Microsoft.Web.Infrastructure" version="1.0.0.0" targetFramework="net461" />
-  <package id="Newtonsoft.Json" version="10.0.3" targetFramework="net461" />
+  <package id="Newtonsoft.Json" version="10.0.2" targetFramework="net461" />
```
Slight version regression here. Probably makes no difference though.

adamralph commented on exercises/02-publish-subscribe/before/Divergent.Finance.API/packages.config:
```diff
@@ -1,28 +1,20 @@
+  <package id="Microsoft.Owin.SelfHost" version="3.1.0" targetFramework="net461" />
   <package id="Microsoft.Web.Infrastructure" version="1.0.0.0" targetFramework="net461" />
-  <package id="Newtonsoft.Json" version="10.0.3" targetFramework="net461" />
+  <package id="Newtonsoft.Json" version="10.0.2" targetFramework="net461" />
```
And here.

adamralph commented on exercises/02-publish-subscribe/before/Divergent.Sales.API/packages.config:
```diff
@@ -1,29 +1,30 @@
-  <package id="Microsoft.AspNet.Mvc" version="5.2.3" targetFramework="net461" />
-  <package id="Microsoft.AspNet.Razor" version="3.2.3" targetFramework="net461" />
+  <package id="Microsoft.AspNet.Mvc" version="5.2.0" targetFramework="net461" />
+  <package id="Microsoft.AspNet.Razor" version="3.2.0" targetFramework="net461" />
```
Slight version regressions here. Do they matter?

adamralph commented on exercises/02-publish-subscribe/before/Divergent.Sales.API/packages.config:
```diff
@@ -1,29 +1,30 @@
-  <package id="Microsoft.AspNet.WebPages" version="3.2.3" targetFramework="net461" />
-  <package id="Microsoft.CodeDom.Providers.DotNetCompilerPlatform" version="1.0.7" targetFramework="net461" />
-  <package id="Microsoft.Net.Compilers" version="2.3.2" targetFramework="net461" developmentDependency="true" />
+  <package id="Microsoft.AspNet.WebPages" version="3.2.0" targetFramework="net461" />
```
And here.

adamralph commented on exercises/02-publish-subscribe/before/Divergent.Sales.API/packages.config:
```diff
@@ -1,29 +1,30 @@
-  <package id="Radical.Bootstrapper.Windsor" version="2.1.1" targetFramework="net461" />
+  <package id="Radical" version="1.5.0.1" targetFramework="net461" />
+  <package id="Radical.Bootstrapper" version="1.0.0.0" targetFramework="net461" />
+  <package id="Radical.Bootstrapper.Windsor" version="1.0.0.0" targetFramework="net461" />
```
And here.

adamralph commented on exercises/02-publish-subscribe/before/Divergent.Sales.API/WebApiConfig.cs:
```diff
@@ -1,33 +1,37 @@
     }
-}
\ No newline at end of file
+}
```
It looks like quite a lot of statements have been re-ordered here. Would it be better to preserve the ordering to simplify the diff?

SzymonPobiega commented:
This is done from implementation point of view. I've smoke tested all the solutions by creating an order and they seem to work. I think I haven't missed anything.

dvdstelt commented:
We really need to have specific reviewers run through the exercises and see if everything still works.

SzymonPobiega commented:
@adamralph I can easily consolidate the json.net usages (revert downgrade) but can't do anything with the 5.2.4 references to some of the asp.net pckages. Tried using 5.2.3 but got an error from nuget:

```
Error downloading 'Microsoft.AspNet.WebApi.OwinSelfHost.5.2.3 : Microsoft.AspNet.WebApi.Owin [5.2.3, 5.3.0), Microsoft.Owin.Host.HttpListener [2.0.2, ), Microsoft.Owin.Hosting [2.0.2, )' from 'https://api.nuget.org/v3-flatcontainer/microsoft.aspnet.webapi.owinselfhost/5.2.3/microsoft.aspnet.webapi.owinselfhost.5.2.3.nupkg'.
  An error occurred while sending the request.
  The underlying connection was closed: An unexpected error occurred on a receive.
  Unable to read data from the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
  A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
```

SzymonPobiega commented:
I updated json to 10.0.3 in all projects/solutions/exercises.

adamralph commented:
@SzymonPobiega thanks, I'll take another look. Unfortunately we have a conflict now. Would you mind rebasing before I review it again? At this point, I would suggest squashing all the commits into one as part of the rebase.

adamralph commented:
@SzymonPobiega FYI I amended the commit with two changes:

- The connection strings had to change from `microservices-workshop` to `particular-workshop` after the merge of https://github.com/Particular/Workshop/pull/239
- The exercise 4 Sales API folders still had a redundant `web.config`

adamralph commented:
Marking as WIP because, for some reason, I keep getting a CORS error when running exercise 1.

adamralph commented:
@SzymonPobiega the CORS error was a red herring. There's something strange going on though. Whenever I run the ex 1 after solution, the service APIs all give 404's, which causes the gateway to return a 500. I don't see this problem with other exercises.

SzymonPobiega commented:
@adamralph I found the problem. The root cause was lack of convention-based routing in the Sales API project. The Sales API controller uses it because it does not specify the route via an attribute.

mauroservienti commented:
Should we move everything to attribute based routing? Since it’s the de facto standard for Asp.Net

adamralph commented:
> Should we move everything to attribute based routing? Since it’s the de facto standard for Asp.Net

We could do that. I don't mind if this happens in this PR or a later one.

adamralph commented:
> Should we move everything to attribute based routing? Since it’s the de facto standard for Asp.Net

@mauroservienti we could do that. I don't mind if this happens in this PR or a later one.

SzymonPobiega commented:
@adamralph @mauroservienti attribute-based routing done. I removed the convention setup code from the `Sales.API` projects and added a non-parametrized `[Route]` attribute to the `Get` method.

DavidBoike commented:
I ran through Exercises 3 and 4. It was slick, everything worked. Much better without all those browser windows popping up!

adamralph commented:
I've run through exercises 1 and 2 and all good.

SzymonPobiega commented:
:tada:
